File: Factorial.java
/**
 * Factorial Implementation
 * Time Complexity (Iterative & Recursive): O(n)
 * Space Complexity: Iterative: O(1) | Recursive: O(n) stack space
 */
public class Factorial {
    static int stepCount = 0;

    public static void factIterative(int n) {
        stepCount = 0;
        long factorial = 1; // Used long to prevent rapid overflow with factorials
        
        for (int i = 1; i <= n; i++) {
            factorial *= i;
            stepCount++;   
        }
        
        System.out.println("Iterative Factorial of " + n + ": " + factorial);
        System.out.println("Step Count: " + stepCount);
    }

    public static long factRecursive(int n) {
        stepCount++;   
        if (n <= 1) {
            return 1;
        }
        return n * factRecursive(n - 1);
    }

    public static void main(String[] args) {
        int n = 5;
        System.out.println("Calculating factorial for n = " + n + "\n");
        
        factIterative(n);
        
        stepCount = 0;
        long result = factRecursive(n);
        System.out.println("Recursive Factorial of " + n + ": " + result);
        System.out.println("Step Count: " + stepCount);
    }
}

File: factorial.cpp
#include <iostream>

using namespace std;

int stepCount = 0;

void factIterative(int n) {
    stepCount = 0;
    long long factorial = 1;

    // Multiply numbers 1, 2, 3 ... n to build n!.
    for (int i = 1; i <= n; i++) {
        factorial *= i;
        stepCount++;
    }

    cout << "Iterative Factorial of " << n << ": " << factorial << "\n";
    cout << "Step Count: " << stepCount << "\n";
}

long long factRecursive(int n) {
    stepCount++;
    // Base case: 0! and 1! are both 1.
    if (n <= 1) {
        return 1;
    }
    // Recursive definition: n! = n * (n-1)!
    return 1LL * n * factRecursive(n - 1);
}

int main() {
    int n = 5;
    cout << "Calculating factorial for n = " << n << "\n\n";

    // Iterative method uses a loop.
    factIterative(n);

    stepCount = 0;
    // Recursive method calls itself until base case is reached.
    long long result = factRecursive(n);
    cout << "Recursive Factorial of " << n << ": " << result << "\n";
    cout << "Step Count: " << stepCount << "\n";

    return 0;
}
